hs/src/Reach/Connector/ALGO.hs

```diff
@@ -3015,9 +3040,29 @@ cmeth sigi = \case
 
 bindFromArgs :: [DLVarLet] -> App a -> App a
 bindFromArgs vs m = do
-  -- XXX deal with >15 args
-  let go (v, i) = sallocVarLet v False (code "txna" ["ApplicationArgs", texty i] >> cfrombs (varLetType v))
-  foldl' (flip go) m (zip vs [(1::Integer) ..])
+  let goSingle (v, i) = sallocVarLet v False (code "txna" ["ApplicationArgs", texty i] >> cfrombs (varLetType v))
+  let goSingles singles k =
+        foldl' (flip goSingle) k (zip singles [(1 :: Integer) ..])
+  -- When there are more than 15 args, the 15th arg is a tuple of args 15+
+  case 15 < (length vs) of
+    False -> do
+      goSingles vs m
+    True -> do
+      -- TODO - I wrote this thinking it was the code path I needed to edit, but it's not exercised by my test.  I should write a test that exercises it and be sure it works.
+      let (vs14, vsMore) = splitAt 14 vs
+      let tupleTy = T_Tuple $ map varLetType vsMore
+      let tupleDupExtract i = do
+            op "dup"
+            cTupleRef (SrcLoc Nothing Nothing Nothing) tupleTy i
+      let goTuple (v, i) = sallocVarLet v False (tupleDupExtract i)
+      let m2 = do
+            -- TODO - instead of this dup/pop dance, I think I can just access arg 15 many times
```

Comment:
`txna AppArg i` costs 3 bytes. `dup`/`pop` are 1. They have cost 1.

(A) What you have is --- 3 + 1n + 1

(B) Referencing it over and over is 3n

(C) If you alloc a slot with `salloc_` then you can read it once (3), store it (2), then reference it (2)... the cost is 3 + 2 + 2n

So:

A: 4+n size; 2+n cost
B: 3n size; n cost
C: 5+2n; 2+n cost

So, I think what you did is optimal.